PMLL.c

```diff
@@ -0,0 +1,102 @@
+/********************************************************************
+  PMLL.c  —  Persistent-Memory Logic Loop (CPU reference)
+  • ring-buffer KV store per attention head
+  • zero external dependencies
+  • thread-safe if you call it from a single OpenMP thread per batch
+  • MIT licence (same as llm.c)
+*********************************************************************/
+
+#include "pmll.h"           /* public ABI */
+#include <stdlib.h>         /* malloc / free */
+#include <string.h>         /* memcpy / memset */
+
+/*---------------------------------------------------------------
+  Life-cycle helpers
+----------------------------------------------------------------*/
+
+/* Allocate and zero-initialise the ring buffers.
+   Return 0 on success, −1 on out-of-memory. */
+int pmll_init(pmll_state *S, int NH, int hs)
+{
+    S->T  = 0;
+    S->hs = hs;
+
+    const size_t bytes = (size_t)NH * MAX_MEM_T * hs * sizeof(float);
+
+    S->k = (float*)calloc(1, bytes);
+    S->v = (float*)calloc(1, bytes);
+
+    return (S->k && S->v) ? 0 : -1;
+}
+
+/* Discard history but keep the allocated buffers. */
+void pmll_reset(pmll_state *S) { S->T = 0; }
+
+/* Free device memory. */
+void pmll_free(pmll_state *S)
+{
+    free(S->k);
+    free(S->v);
+    S->k = S->v = NULL;
+    S->T = S->hs = 0;
+}
+
+/*---------------------------------------------------------------
+  Data flow
+----------------------------------------------------------------*/
+
+/* Read historic KV for head *h* to caller-provided scratch.
+   out_k/out_v length = (Tmem + Tctx) * hs floats. */
+void pmll_read(float *out_k, float *out_v,
```

Comment:
Unless there is a specific reason not to do so, `pmll_read()` should have the 
`const pmll_state *S` argument first for consistency.